Lab1/Lab1/Program.cs

```diff
@@ -0,0 +1,145 @@
+                StartPage();
+            }
+        }
+
+        private static void ReceiveMessages()
+        { 
+            try
+            {
+                while (true)
+                {
+                    StringBuilder value = new ();
+                    byte[] data = new byte[256];
+                    do
+                    {
+                        var bytes = Socket.ReceiveFrom(data, ref _remotePoint);
+                        value.Append(Encoding.Unicode.GetString(data, 0, bytes));
+                    }
+                    while (Socket.Available > 0);
+
+                    switch (value[0])
+                    {
+                        case '/':
+                        {
+                            value.Remove(0,1);
+                            if (Convert.ToInt32(value.ToString()) != Messages.Count)
+                            {
+                                Error("Some message was not delivered to you.");
+                                Socket.Send(Encoding.Unicode.GetBytes("|"));
+                            }
+
+                            break;
+                        }
+                        case '|':
+                            var mes = Encoding.Unicode.GetBytes(Messages[-1].Item1 + "|"+Messages[-1].Item2);
+                            Socket.SendTo(mes, _remotePoint);
+                            Error("Some message was not delivered to your companion.");
+                            break;
+                        default:
+                            string temp = "";
+                            for (var i = 0; i < value.Length; i++)
+                            {
+                                if (value[i] != '|')
+                                {
+                                    temp += value[i];
+                                }
+                                else
+                                {
+                                    value.Remove(0, i + 1);
+                                    Messages.Add((Convert.ToDateTime(temp), value.ToString()));
+                                    DisplayMessageHistory();
+                                    break;
+                                }
+                            }
+                            break;
+                    }
+                }
+            }
+            catch(Exception ex)
+            {
+                Error(ex.Message);
+            }
+        }
+
+        private static void SendMessage()
+        {
+            while (true)
+            {
+                var message = Console.ReadLine();
+                var time = DateTime.UtcNow;
+                Messages.Add((time, _nick + ": " + message));
+                DisplayMessageHistory();
+                var data = Encoding.Unicode.GetBytes(time + "|"+_nick + ": " + message);
+                Socket.SendTo(data, _remotePoint);
+                Thread.Sleep(500);
+                Socket.SendTo(Encoding.Unicode.GetBytes("/" + Messages.Count),_remotePoint);
```

Comment:
So you send message count after the actual message so that the other client could verify if nothing is lost, right?

Wound't it be better, to assign ID to the message packet? In this case the receiver could even re-request lost messages if it discovers a gap in ID sequence.

Reply:
@gmltA Yeah, that's right, but i have functionality, that you have described. When other client see that message count doesn't match it will send message that it didn't get last message and ask for resend of the last message. Probably yeah it will be better to create id instead of DateTime and message counter, because id can be order-related parameter and control messages that have been lost at the same time